File: Source/description/description.cpp
#include <type_traits>
#include <iterator>

#include "meta/meta.h"
#include "path/svg/parse.h"

#include "description/description.h"

namespace rvg {
    namespace description {

using rvg::color::unorm_to_uint8_t;
using rvg::scene::Scene;

struct Depths { int blur, fade, clip, xform; };

static void build_stencil(const std::vector<Stencil> &stencil, Depths &depths,
    Scene &forward);

static void build_painted(const std::vector<Painted> &painted, Depths &depths,
    Scene &forward);

Shape path(const char *svg) {
    auto p = std::make_shared<path::Path>();
    if (!path::svg::tokens_to_instructions(svg, *p))
        p->clear();
    return Shape(p);
}

Shape circle(float x, float y, float r) {
    return Shape(std::make_shared<shape::Circle>(x, y, r));
}

Shape triangle(float x1, float y1, float x2, float y2, float x3, float y3) {
    return Shape(std::make_shared<shape::Triangle>(x1, y1, x2, y2, x3, y3));
}

Shape rect(float x, float y, float width, float height) {
    return Shape(std::make_shared<shape::Rect>(x, y, width, height));
}

Shape polygon(const std::initializer_list<float> &coordinates) {
    return Shape(std::make_shared<shape::Polygon>(coordinates));
}

Shape path(const ILToken &svg) {
    auto p = std::make_shared<path::Path>();
    if (!path::svg::tokens_to_instructions(svg, *p))
        p->clear();
    return Shape(p);
}

Painted nzfill(const Shape &shape, const Paint &paint) {
    return make_painted_primitive(scene::WindingRule::non_zero, shape, paint);
}

Painted zfill(const Shape &shape, const Paint &paint) {
    return make_painted_primitive(scene::WindingRule::zero, shape, paint);
}

Painted ofill(const Shape &shape, const Paint &paint) {
    return make_painted_primitive(scene::WindingRule::odd, shape, paint);
}

Painted efill(const Shape &shape, const Paint &paint) {
    return make_painted_primitive(scene::WindingRule::even, shape, paint);
}

Painted fill(const Shape &shape, const Paint &paint) {
    return make_painted_primitive(scene::WindingRule::non_zero, shape, paint);
}

Painted eofill(const Shape &shape, const Paint &paint) {
    return make_painted_primitive(scene::WindingRule::odd, shape, paint);
}

Painted fade(uint8_t opacity, const VPainted &painted) {
    return make_painted_faded(opacity, painted);
}

Painted fade(uint8_t opacity, const ILPainted &painted) {
    return make_painted_faded(opacity, painted);
}

Painted transform(const Xform &xf, const VPainted &painted) {
    return make_painted_xformed(xf, painted);
}

Painted transform(const Xform &xf, const ILPainted &painted) {
    return make_painted_xformed(xf, painted);
}

Painted blur(float radius, const VPainted &painted) {
    return make_painted_blurred(radius, painted);
}

Painted blur(float radius, const ILPainted &painted) {
    return make_painted_blurred(radius, painted);
}

Painted blur(float radius, const Painted &painted) {
    return blur(radius, { painted });
}

Stencil nzstencil(const Shape &shape) {
    return make_stencil_primitive(scene::WindingRule::non_zero, shape);
}

Stencil zstencil(const Shape &shape) {
    return make_stencil_primitive(scene::WindingRule::zero, shape);
}

Stencil ostencil(const Shape &shape) {
    return make_stencil_primitive(scene::WindingRule::odd, shape);
}

Stencil estencil(const Shape &shape) {
    return make_stencil_primitive(scene::WindingRule::even, shape);
}

Stencil stencil(const Shape &shape) {
    return make_stencil_primitive(scene::WindingRule::non_zero, shape);
}

Stencil eostencil(const Shape &shape) {
    return make_stencil_primitive(scene::WindingRule::odd, shape);
}

Painted clip(const VStencil &clipper, const VPainted &clippee) {
    return make_painted_clipped(clipper, clippee);
}

Painted clip(const ILStencil &clipper, const ILPainted &clippee) {
    return make_painted_clipped(clipper, clippee);
}

Painted clip(const Stencil &clipper, const ILPainted &clippee) {
    return clip({ clipper }, clippee);
}

Painted clip(const ILStencil &clipper, const Painted &clippee) {
    return clip(clipper, { clippee });
}

Painted clip(const Stencil &clipper, const Painted &clippee) {
    return clip({ clipper }, { clippee });
}

Stencil clip(const ILStencil &clipper, const ILStencil &clippee) {
    return make_stencil_clipped(clipper, clippee);
}

Stencil clip(const VStencil &clipper, const VStencil &clippee) {
    return make_stencil_clipped(clipper, clippee);
}

Stencil clip(const Stencil &clipper, const ILStencil &clippee) {
    return clip({ clipper }, clippee);
}

Stencil clip(const ILStencil &clipper, const Stencil &clippee) {
    return clip(clipper, { clippee });
}

Stencil clip(const Stencil &clipper, const Stencil &clippee) {
    return clip({ clipper }, { clippee });
}

Stencil transform(const Xform &xf, const ILStencil &stencil) {
    return make_stencil_xformed(xf, stencil);
}

Stencil transform(const Xform &xf, const VStencil &stencil) {
    return make_stencil_xformed(xf, stencil);
}

template <typename IT>
typename std::enable_if<std::is_same<
    typename std::iterator_traits<IT>::value_type, Stencil>::value>::type
build_stencil(IT first, IT last, Depths &depths, Scene &forward) {
    using ST = Stencil::Type;
    for ( ; first != last; ++first) {
        const auto &st = *first;
        if (st.type() == ST::primitive) {
            const auto &prim = st.primitive();
            forward.stencil_element(
                prim.winding_rule(),
                prim.shape().transformed(st.xf()));
        } else {
            // all stencil objects have their own xf
            // if the xf is not the identity, we bracket
            // around it with a transform bracket
            bool xformed = !st.xf().is_identity();
            if (xformed) {
                forward.begin_transform(depths.xform, st.xf());
                ++depths.xform;
            }
            switch (st.type()) {
                case ST::xformed:
                    // already bracketed outside
                    build_stencil(st.xformed().stencil(), depths, forward);
                    break;
                case ST::clipped: {
                    const auto &clipped = st.clipped();
                    forward.begin_clip(static_cast<uint16_t>(depths.clip++));
                    build_stencil(clipped.clipper(), depths, forward);
                    forward.activate_clip(static_cast<uint16_t>(depths.clip-1));
                    build_stencil(clipped.clippee(), depths, forward);
                    forward.end_clip(static_cast<uint16_t>(--depths.clip));
                    break;
                }
                default:
                    // do nothing
                    break;
            }
            if (xformed) {
                --depths.xform;
                forward.end_transform(depths.xform, st.xf());
            }
        }
    }
}

static void build_stencil(const std::vector<Stencil> &stencil, Depths &depths,
    Scene &forward) {
    build_stencil(stencil.begin(), stencil.end(), depths, forward);
}

template <typename IT>
typename std::enable_if<std::is_same<
    typename std::iterator_traits<IT>::value_type, Painted>::value>::type
build_painted(IT first, IT last, Depths &depths, Scene &forward) {
    using PT = Painted::Type;
    for ( ; first != last; ++first) {
        const Painted &p = *first;
        if (p.type() == PT::primitive) {
            const auto &prim = p.primitive();
            forward.painted_element(prim.winding_rule(),
                prim.shape().transformed(p.xf()),
                prim.paint().transformed(p.xf()));
        } else {
            // all painted objects have their own xf
            // if the xf is not the identity, we bracket
            // around it with a transform bracket
            bool xformed = !p.xf().is_identity();
            if (xformed) {
                forward.begin_transform(depths.xform, p.xf());
                ++depths.xform;
            }
            switch (p.type()) {
                case PT::faded: {
                    const auto &faded = p.faded();
                    forward.begin_fade(static_cast<uint16_t>(depths.fade++),
                        faded.opacity());
                    build_painted(faded.painted(), depths, forward);
                    forward.end_fade(static_cast<uint16_t>(--depths.fade),
                        faded.opacity());
                    break;
                }
                case PT::xformed: {
                    // already bracketed outside
                    build_painted(p.xformed().painted(), depths, forward);
                    break;
                }
                case PT::blurred: {
                    const auto &blurred = p.blurred();
                    forward.begin_blur(static_cast<uint16_t>(depths.blur++),
                        blurred.radius());
                    build_painted(blurred.painted(), depths, forward);
                    forward.end_blur(static_cast<uint16_t>(--depths.blur),
                        blurred.radius());
                    break;
                }
                case PT::clipped: {
                    const auto &clipped = p.clipped();
                    forward.begin_clip(static_cast<uint16_t>(depths.clip++));
                    build_stencil(clipped.clipper(), depths, forward);
                    forward.activate_clip(static_cast<uint16_t>(depths.clip-1));
                    build_painted(clipped.clippee(), depths, forward);
                    forward.end_clip(static_cast<uint16_t>(--depths.clip));
                    break;
                }
                default:
                    // do nothing
                    break;
            }
            if (xformed) {
                --depths.xform;
                forward.end_transform(depths.xform, p.xf());
            }
        }
    }
}

static void build_painted(const std::vector<Painted> &painted, Depths &depths,
    Scene &forward) {
    build_painted(painted.begin(), painted.end(), depths, forward);
}

template <typename IT>
typename std::enable_if<std::is_same<
    typename std::iterator_traits<IT>::value_type, Painted>::value,
XformableScene>::type scene(IT first, IT last) {
    scene::ScenePtr s = std::make_shared<scene::Scene>();
    Depths depths{0,0,0,0};
    build_painted(first, last, depths, *s);
    return XformableScene(s);
}

XformableScene scene(const ILPainted &painted) {
    return scene(painted.begin(), painted.end());
}

XformableScene scene(const std::vector<Painted> &painted) {
    return scene(painted.begin(), painted.end());
}

XformableScene scene(const Painted &painted) {
    return scene({ painted });
}

Paint solid_color(const rvg::color::RGBA8 &color, uint8_t opacity) {
    return Paint(color, opacity);
}

RampPtr ramp(Spread spread, const std::initializer_list<Stop> &stops) {
    return std::make_shared<paint::Ramp>(spread, stops);
}

RampPtr ramp(const std::initializer_list<Stop> &stops) {
    return std::make_shared<paint::Ramp>(Spread::pad, stops);
}

Paint texture(Spread spread, const IImagePtr &image_ptr, uint8_t opacity) {
    return Paint(std::make_shared<paint::Texture>(spread, image_ptr), opacity);
}

Paint texture(const IImagePtr &image_ptr, uint8_t opacity) {
    return Paint(std::make_shared<paint::Texture>(Spread::pad, image_ptr),
        opacity);
}

Paint linear_gradient(const RampPtr &ramp_ptr, float x1, float y1, float x2,
    float y2, uint8_t opacity) {
    return Paint(std::make_shared<paint::LinearGradient>(ramp_ptr,
        x1, y1, x2, y2), opacity);
}

Paint radial_gradient(const RampPtr &ramp_ptr, float cx, float cy, float fx,
    float fy, float r, uint8_t opacity) {
    return Paint(std::make_shared<paint::RadialGradient>(ramp_ptr,
        cx, cy, fx, fy, r), opacity);
}

Window window(float xl, float yb, float xr, float yt) {
    return Window(xl, yb, xr, yt);
}

Viewport viewport(int xl, int yb, int xr, int yt) {
    return Viewport(xl, yb, xr, yt);
}

namespace cap {
    stroke::Cap butt = stroke::Cap::butt;
    stroke::Cap round = stroke::Cap::round;
    stroke::Cap square = stroke::Cap::square;
    stroke::Cap triangle = stroke::Cap::triangle;
    stroke::Cap fletching = stroke::Cap::fletching;
}

namespace join {
    stroke::Join arcs = stroke::Join::arcs;
    stroke::Join miter = stroke::Join::miter;
    stroke::Join miterclip = stroke::Join::miterclip;
    stroke::Join round = stroke::Join::round;
    stroke::Join bevel = stroke::Join::bevel;
}

namespace method {
    stroke::Method curves = stroke::Method::curves;
    stroke::Method lines = stroke::Method::lines;
    stroke::Method driver = stroke::Method::driver;
}

namespace spread {
    paint::Spread repeat = paint::Spread::repeat;
    paint::Spread reflect = paint::Spread::reflect;
    paint::Spread pad = paint::Spread::pad;
    paint::Spread transparent = paint::Spread::transparent;
}

} } // namespace rvg::description

File: Source/description.lua
local _M = {}

local scene = require"scene"
local color = require"color"
local paint = require"paint"
local strokestyle = require"strokestyle"
local xform = require"xform"
local xformable = require"xformable"
local svd = require"svd"

local shape_types = {
    triangle = true,
    path = true,
    circle = true,
    rect = true,
    stroked = true,
    polygon = true,
}

local function is_shape(s)
    return shape_types[s.type]
end

local painted_meta = { __index = {} }
local stencil_meta = { __index = {} }

function new_primitive_painted(winding_rule)
    return function(s, p)
        assert(is_shape(s), "expected shape")
        if color.is_color(p) then
            p = paint.solid_color(p)
        end
        assert(paint.is_paint(p), "expected paint")
        return setmetatable({
            type = "primitive",
            winding_rule = winding_rule,
            shape = s,
            paint = p,
            xf = xform.identity()
        }, painted_meta)
    end
end

_M.nzfill = new_primitive_painted("non-zero")
_M.fill = _M.nzfill
_M.zfill = new_primitive_painted("zero")
_M.ofill = new_primitive_painted("odd")
_M.eofill = _M.ofill
_M.efill = new_primitive_painted("even")

function new_primitive_stencil(winding_rule)
    return function(s)
        assert(is_shape(s), "expected shape")
        return setmetatable({
            type = "primitive",
            winding_rule = winding_rule,
            shape = s,
            xf = xform.identity()
        }, stencil_meta)
    end
end

_M.nzstencil = new_primitive_stencil("non-zero")
_M.stencil = _M.nzstencil
_M.zstencil = new_primitive_stencil("zero")
_M.ostencil = new_primitive_stencil("odd")
_M.eostencil = _M.ostencil
_M.estencil = new_primitive_stencil("even")

local function is_painted(p)
    return getmetatable(p) == painted_meta
end

local function is_painted_or_list(p)
    if is_painted(p) then
        return true
    else
        if type(p) ~= "table" or #p < 1 then
            return false
        end
        for i,v in ipairs(p) do
            if not is_painted(v) then
                return false
            end
        end
    end
    return true
end

local function is_stencil(s)
    return getmetatable(s) == stencil_meta
end

local function is_stencil_or_list(s)
    if is_stencil(s) then
        return true
    else
        if type(s) ~= "table" or #s < 1 then
            return false
        end
        for i,v in ipairs(s) do
            if not is_stencil(v) then
                return false
            end
        end
    end
    return true
end

function _M.blur(radius, painted)
    if is_painted_or_list(painted) then
        return setmetatable({
            type = "blurred",
            radius = radius,
            painted = painted,
            xf = xform.identity()
        }, painted_meta)
    else
        error("expected painted")
    end
end

function _M.fade(opacity, painted)
    if is_painted_or_list(painted) then
        return setmetatable({
            type = "faded",
            opacity = opacity,
            painted = painted,
            xf = xform.identity()
        }, painted_meta)
    else
        error("expected painted")
    end
end

function _M.transform(xf, p_or_s)
    if is_stencil_or_list(p_or_s) then
        return setmetatable({
            type = "xformed",
            stencil = p_or_s,
            xf = xf
        }, stencil_meta)
    elseif is_painted_or_list(p_or_s) then
        return setmetatable({
            type = "xformed",
            painted = p_or_s,
            xf = xf
        }, painted_meta)
    else
        error("expected painted or stencil")
    end
end

function _M.clip(clipper, clippee)
    if not is_stencil_or_list(clipper) then
        error("expected stencil clipper")
    end
    if is_stencil_or_list(clippee) then
        return setmetatable({
            type = "clipped",
            clipper = clipper,
            clippee = clippee,
            xf = xform.identity()
        }, stencil_meta)
    elseif is_painted_or_list(clippee) then
        return setmetatable({
            type = "clipped",
            clipper = clipper,
            clippee = clippee,
            xf = xform.identity()
        }, painted_meta)
    else
        error("expected painted or stencil clippee")
    end
end

local function build_stencil(stencil_list, depths, forward)
    if is_stencil(stencil_list) then
        stencil_list = { stencil_list }
    end
    for i,st in ipairs(stencil_list) do
        assert(is_stencil(st), "expected stencil")
        if st.type == "primitive" then
            forward:stencil_element(
                st.winding_rule,
                st.shape:transformed(st.xf))
        else
            -- all stencil objects have their own xf
            -- if the xf is not the identity, we bracket
            -- around it with a transform bracket
            local xformed = not st.xf:is_identity()
            if xformed then
                forward:begin_transform(depths.xform, st.xf)
                depths.xform = depths.xform + 1
            end
            if st.type == "xformed" then
                -- already bracketed outside
                build_stencil(st.stencil, depths, forward)
            elseif st.type == "clipped" then
                forward:begin_clip(depths.clip)
                depths.clip = depths.clip + 1
                build_stencil(st.clipper, depths, forward)
                forward:activate_clip(depths.clip-1)
                build_stencil(st.clippee, depths, forward)
                depths.clip = depths.clip - 1
                forward:end_clip(depths.clip)
            end
            if xformed then
                depths.xform = depths.xform - 1
                forward:end_transform(depths.xform, st.xf)
            end
        end
    end
end

local function build_painted(painted_list, depths, forward)
    if is_painted(painted_list) then
        painted_list = { painted_list }
    end
    for i, p in ipairs(painted_list) do
        assert(is_painted(p), "expected painted")
        if p.type == "primitive" then
            forward:painted_element(
                p.winding_rule,
                p.shape:transformed(p.xf),
                p.paint:transformed(p.xf))
        else
            local xformed = not p.xf:is_identity()
            if xformed then
                forward:begin_transform(depths.xform, p.xf)
                depths.xform = depths.xform + 1
            end
            if p.type == "faded" then
                forward:begin_fade(depths.fade, p.opacity)
                depths.fade = depths.fade + 1
                build_painted(p.painted, depths, forward)
                depths.fade = depths.fade - 1
                forward:end_fade(depths.fade, p.opacity)
            elseif p.type == "xformed" then
                -- already brackted outside
                build_painted(p.painted, depths, forward)
            elseif p.type == "blurred" then
                forward:begin_blur(depths.blur, p.radius)
                depths.blur = depths.blur + 1
                build_painted(p.painted, depths, forward)
                depths.blur = depths.blur - 1
                forward:end_blur(depths.blur, p.radius)
            elseif p.type == "clipped" then
                forward:begin_clip(depths.clip)
                depths.clip = depths.clip + 1
                build_stencil(p.clipper, depths, forward)
                forward:activate_clip(depths.clip-1)
                build_painted(p.clippee, depths, forward)
                depths.clip = depths.clip - 1
                forward:end_clip(depths.clip)
            end
            if xformed then
                depths.xform = depths.xform - 1
                forward:end_transform(depths.xform, p.xf)
            end
        end
    end
end

function _M.scene(painted_list)
    local forward = scene.scene()
    local depths = { blur = 1, fade = 1, clip = 1, xform = 1 }
    build_painted(painted_list, depths, forward)
    return forward
end

function _M.stroke(width_or_style)
    return strokestyle.default():stroked(width_or_style)
end

xformable.setmethods(painted_meta.__index)
xformable.setmethods(stencil_meta.__index)

return _M
